junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java

```diff
@@ -70,17 +91,116 @@ public void execute(ExecutionRequest request) {
+	protected interface CloseableExecutor extends Executor, AutoCloseable {
+		default public void close() {
+		};
+	}
+
+	/**
+	 * {@link CloseableExecutor} that executes tasks synchronously.
+	 *
+	 * @since 5.8
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected class SerialExecutor implements CloseableExecutor {
+		public void execute(Runnable command) {
+			command.run();
+		}
+
+	}
+
+	/**
+	 * {@link CloseableExecutor} backed by a {@link ForkJoinPool} that executes tasks asynchronously
+	 * based on the settings in {@value #PARALLEL_CONFIG}. Clients *must* implement their own logic
+	 * to wait for tasks to complete.
+	 *
+	 * @since 5.8
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected class ParallelExecutor implements CloseableExecutor {
+
+		private final ExecutorService executorService;
+
+		/**
+		 * @param parameters test execution configuration
+		 */
+		public ParallelExecutor(ConfigurationParameters parameters) {
+			ConfigurationParameters prefixedParameters = new PrefixedConfigurationParameters(parameters,
+				PARALLEL_CONFIG);
+			String strategyName = prefixedParameters.get("strategy").orElse("dynamic");
+			ParallelExecutionConfigurationStrategy executionStrategy = DefaultParallelExecutionConfigurationStrategy.valueOf(
+				strategyName.toUpperCase());
+			ParallelExecutionConfiguration executionConfiguration = executionStrategy.createConfiguration(
+				prefixedParameters);
+			executorService = createExecutorService(executionConfiguration);
+		}
+
+		public void execute(Runnable command) {
+			executorService.execute(command);
+		}
+
+		public void close() {
+			executorService.shutdown();
+		}
+
+		/**
+		 * Create an executor service based on the parallel configuration properties. This emulates the logic in {@link org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService ForkJoinPoolHierarchicalTestExecutorService}.
+		 */
+		protected ExecutorService createExecutorService(ParallelExecutionConfiguration configuration) {
+			return Try.call(() -> {
+				// Try to use constructor available in Java >= 9
+				Constructor<ForkJoinPool> constructor = ForkJoinPool.class.getDeclaredConstructor(Integer.TYPE,
+					ForkJoinWorkerThreadFactory.class, UncaughtExceptionHandler.class, Boolean.TYPE, Integer.TYPE,
+					Integer.TYPE, Integer.TYPE, Predicate.class, Long.TYPE, TimeUnit.class);
+				return constructor.newInstance(configuration.getParallelism(),
+					ForkJoinPool.defaultForkJoinWorkerThreadFactory, null, false, configuration.getCorePoolSize(),
+					configuration.getMaxPoolSize(), configuration.getMinimumRunnable(), null,
+					configuration.getKeepAliveSeconds(), TimeUnit.SECONDS);
+			}).orElseTry(() -> {
+				// Fallback for Java 8
+				return new ForkJoinPool(configuration.getParallelism(), ForkJoinPool.defaultForkJoinWorkerThreadFactory,
+					null, false);
+			}).getOrThrow(cause -> new JUnitException("Failed to create ForkJoinPool", cause));
+		}
+
 	}
```

Comment:
Let's introduce a shared factory for creating a `ForkJoinPool` in `org.junit.platform.engine.support.hierarchical` and use it for both Jupiter and Vintage.

Reply:
This is done. Now both the Jupiter and Vintage engines use the custom thread factory.